### PR TITLE
[WIP] Py 2-3 guide

### LINKED ISSUE #4855: Python 2/3 migration guide
# User story

As a {developer} working on Py2/3 migration for CKAN, I want a clear migration guide, so I can provide Py2/3 compatibility in a consistent way using a common set of tools.

# Use cases

- Contributing Py2/3 compatible code in CKAN Core
- Contributing Py2/3 compatible code in a CKAN Extension
- Starting a new CKAN Extension with Py2/3 compatibility

### PR DESCRIPTION
**Pull request by pwalsh**
Fixes #4855 

### Proposed fixes:

- Guide for developers contributing to Py2-3 codebases around CKAN.

### CODE REVIEW

**Comment by pwalsh** [comment]
@pdelboca @amercader 

Specially on Flask-based APIs for extensions, can @pdelboca contribute for this section, as I am not aware of the details, and you are both working on a codebase for this right now.

**Comment by pdelboca** [comment]
Great to have this documented! There are some pieces of information in several tickets, and a Issue for Py 2-3 Guide for extensions (https://github.com/ckan/ckan/issues/4792)

I will leave here some of those interesting pieces of information that are suitable to be written down and merged in this PR:
- In https://github.com/ckan/ckan/issues/4791 there is a good approach on how to think the migration of extensions to Flask.
- There are already some core views migrated to Flask that can be use as reference and inspiration to design extensions (https://github.com/ckan/ckan/tree/master/ckan/views)
- There are some extensions already made in Flask like for example: https://github.com/okfn/ckanext-collaborators 
- There is also an old but good guide and documentation for the migration with lot of useful information in the Wiki: https://github.com/ckan/ckan/wiki/Migration-from-Pylons-to-Flask
- Not sure if related but it is worth noticing that there is ongoing work to migrate the command line tool (https://github.com/ckan/ckan/issues/4639) as it may also have an impact on how to design commands for the extensions.

File: doc/contributing/py-2-3-guide.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+Ensure you read the references below for further background information on all these tips.
+
+Exception handling
+##################
+
+Python-Future provides a clean interface for exception handling compatibility::
+
+  from future.utils import raise_
+  traceback = sys.exc_info()[2]
+  raise_(ValueError, "dodgy value", traceback)
```
Comment (wardi) [inline doc/contributing/py-2-3-guide.rst]: -1 on python-future. I strongly dislike rewriting exceptions this way. I think the cases in ckan core where we raise exceptions that won't work in python 2/3 should just be fixed (if there are any)

File: doc/contributing/py-2-3-guide.rst [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+  heights = dict(Fred=175, Anne=166, Joe=192)
+
+  # value lists
+  valuelist = listvalues(heights)
+  valuelist = list(itervalues(heights))
+
+  # item lists
+  itemlist = listitems(heights)
+  itemlist = list(iteritems(heights))
```
Comment (wardi) [inline doc/contributing/py-2-3-guide.rst]: I don't see list and dict iteration differences in Python 3 as a big problem. For core we can just assume iterating will be lazy and pass to `list()` when we actually need a list.

File: doc/contributing/py-2-3-guide.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+Future imports
+##############
+
+Import the following at the top of all modules::
+
+  from __future__ import absolute_import
+  from __future__ import division
+  from __future__ import print_function
+  from __future__ import unicode_literals
```
Comment (wardi) [inline doc/contributing/py-2-3-guide.rst]: :+1: from me

File: doc/contributing/py-2-3-guide.rst (hunk 1 above)
Comment (pwalsh) [reply doc/contributing/py-2-3-guide.rst]: Hey, thanks.

As far as I know, raising exceptions with a traceback, and also reprising exceptions, across both 2 and 3 is only possible like this. Of course the standard approach of `Exception as e` is better for simple cases.

Some more info [here](https://python-future.org/compatible_idioms.html#raising-exceptions) and [here](https://franklingu.github.io/programming/2016/06/30/properly-reraise-exception-in-python/).

File: doc/contributing/py-2-3-guide.rst (hunk 2 above)
Comment (pwalsh) [reply doc/contributing/py-2-3-guide.rst]: I guess the issue is, the same syntax in Python 2 and Python 3 has different behaviour, and that make lead to either different behaviours or inconsistencies (or, not).

I'd only suggest such an approach for Py 2-3 codebases, and of course would also prefer the Py 3 only approach of lazy iteration being assumed.

Overall, fine by me if the core team decides iteration differences are not critical, and I can just update the doc to indicate that as a potential gotcha.

File: doc/contributing/py-2-3-guide.rst [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+- Contributing Python 2/3 compatible code in CKAN Core
+- Contributing Python 2/3 compatible code in a CKAN Extension
+- Starting a new CKAN Extension with Python 2/3 compatibility
+
+--------
+Python 3
+--------
+
+From CKAN 2.9, CKAN will officially support Python 2 (≥ 2.6) and Python 3 (≥ 3.6).
```
Comment (amercader) [inline doc/contributing/py-2-3-guide.rst]: CKAN 2.9 will support Python 2 (≥ 2.7) and Python 3 (≥ 3.6).

CKAN 3.0 will only support Python 3 (≥ 3.6)

File: doc/contributing/py-2-3-guide.rst (hunk 3 above)
Comment (amercader) [reply doc/contributing/py-2-3-guide.rst]: Shouldn't we add only the ones needed? I guess it makes sense `unicode_literals` is always present but for instance the print function is only needed in the cli module. I only could find divisions on 5 modules. And regarding the imports, the current codebase favours absolute imports anyway so perhaps we should recommend that and not load `absolute_import`

File: doc/contributing/py-2-3-guide.rst (hunk 1 above)
Comment (amercader) [reply doc/contributing/py-2-3-guide.rst]: I agree this seems overkill. I grepped the code for exceptions that attach tracebacks and couldn't find any (maybe some on the search index rebuild CLI). There are several instances where we catch exceptions and we reraise custom ones like, but most are at the validators and logic level and are later catch by the blueprints.

I think is best to deal with problematic cases as we find them

**Comment by amercader** [comment]
What are people's views on imports from `urllib`? On https://github.com/ckan/ckan/pull/4045 we introduced the six based syntax in cli.py:

```python
from six.moves.urllib.error import HTTPError
from six.moves.urllib.parse import urljoin, urlparse
from six.moves.urllib.request import urlopen
```
I quite like it and I don't think the [alternatives](https://python-future.org/compatible_idioms.html#urllib-module) offer a bigger benefit.

What I would suggest is restrict the `urllib` usage to utilities like `urlparse` and `urlencode` and use requests for actual calls and exception handling.

Does this sound good?

**Comment by davidread** [comment]
@amercader yeah I like this six syntax the best

**Comment by amercader** [comment]
Closing in favour of https://github.com/ckan/ckan/issues/4792. End result: https://github.com/ckan/ckan/wiki/Python-3-migration-guide-for-extensions

**Pull request closed without merging**